lib/elixir/test/elixir/exception_test.exs

```diff
@@ -954,6 +954,18 @@ defmodule ExceptionTest do
     test "ErlangError" do
       assert %ErlangError{original: :sample} |> message == "Erlang error: :sample"
     end
+
+    test "MissingApplicationsError" do
+      assert %MissingApplicationsError{
+               apps: [{:logger, "~> 1.18"}, {:ex_unit, Version.parse_requirement!(">= 0.0.0")}]
+             }
+             |> message == """
+             missing applications:
+
+               * :logger (~> 1.18)
+               * :ex_unit (>= 0.0.0)
+             """
```

Comment:
Here is my suggestion for the raising:

```
unless Application.spec(:foo_bar, :vsn) do
  raise MissingApplicationError,
        description: "application :foobar is required for feature xyz",
        apps: [{:foobar, "~> 1.0"}]
end
```

and it should raise:

```
** (MissingApplicationError) application :foobar is required for feature xyz
To address this, include this application as your dependency:

    {:foobar, "~> 1.0"}

```

I would keep them as tuples, because it is easier to copy and paste.

Reply:
> I would keep them as tuples, because it is easier to copy and paste.

Good call.